rustc: build with system llvm and jemalloc

symphorien opened the pull request:
###### Motivation for this change

We can use our llvm instead of the bundled one, just like fedora does: https://src.fedoraproject.org/rpms/rust/blob/master/f/rust.spec

The closure size goes down a little: 780MB -> 740MB, but now part of this closure is llvm and can be shared. This also benefits compilation time.

The commit also includes some cleanup of tests which no longer exist and don't need to be removed.

I tested compilation of alacritty, firefox, thunderbird and ripgrep on x86_64-linux. (on top of master, 9f8828286212d8ce7b0fb3578c5f209c434dc60a)

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

symphorien committed:
rustc: build with system llvm and jemalloc

Just like fedora does: https://src.fedoraproject.org/rpms/rust/blob/master/f/rust.spec

Also some cleanup of tests which were removed but no longer exist.

Mic92 commented on pkgs/development/compilers/rust/default.nix:
```diff
@@ -29,7 +29,7 @@ in rec {
     ];
 
-    forceBundledLLVM = true;
+    withBundledLLVM = false;
```
@LnL7 @dtzWill anyone who does remember why we switched away from this option? I thing this option was true at some point.

Mic92 replied:
It broke the tests back then https://github.com/NixOS/nixpkgs/commit/7bd191df6a861e02d5bf2b36a03f6dc70e3e6bfd
It could be still ok though:

- Archlinux does this as well: https://git.archlinux.org/svntogit/community.git/tree/trunk/PKGBUILD?h=packages/rust 
- Ubuntu uses the bundled one: https://git.launchpad.net/ubuntu/+source/rustc/tree/debian/control?h=ubuntu/cosmic
- Opensuse does: https://build.opensuse.org/package/view_file/devel:languages:rust/rust/rust.spec?expand=1

Since you deal with rust backports sometimes for firefox what would you prefer @andir ?
I guess in the worst case it should be still possible to just re-enable the bundled version in backported version.

andir replied:
I really have no strong opinion here. I would certainly like to avoid a situation where we have to port LLVM whenever we have to back port rustc for Thunderbird/Firefox.

How much is the speedup of compilation time? If it is significant I could see this being a good thing in general.

symphorien replied:
Compiling rustc with the bundled llvm takes about 1h05 for llvm and 55 min for rustc. So this halves compilation time.
Note that fedora disables codegen tests when using the bundled llvm.

Mic92 replied:
At the moment we do not run tests anyway.

LnL7 replied:
Is everything from their llvm fork upstream now? That was the main reason for this IIRC.

Mic92 replied:
They only seem to merge pull request from changes that have been before in llvm.

LnL7 replied:
Great, I tried to split this out into 2 separate derivations at some point but not rebuilding llvm is even better.

dtzWill commented:
No clue offhand, does `git blame` (or similar) give any explanation?

On Wed, 31 Oct 2018 18:04:32 -0700, Jörg Thalheim <notifications@github.com> wrote:
> Mic92 commented on this pull request.
> 
> 
> 
> > @@ -29,7 +29,7 @@ in rec {
>        ./patches/disable-test-inherit-env.patch
>      ];
>  
> -    forceBundledLLVM = true;
> +    withBundledLLVM = false;
> 
> @LnL7 @dtzWill anyone who does remember why we switched away from this option? I thing this option was true at some point.
> 
> -- 
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/NixOS/nixpkgs/pull/49557#pullrequestreview-170526432Non-text part: text/html

Mic92 commented:
I will test the aarch64 build. I also would like to see macOS tested.

Mic92 commented:
@periklis It would be interesting to see if this pull request would solve #49708 in a more general way on macOS.

periklis commented:
@Mic92 I started a build on my darwin builder. I will report the results when done

Mic92 commented:
@danieldk maybe you can also have look at this w.r.t. to macOS.

danieldk commented:
Does not get very far on Darwin (Mojave, latest Xcode):

https://gist.githubusercontent.com/danieldk/32e8d9beba4c5d7e442f917a06ab0635/raw/43f164906002dfa3b1b49c04e33d19ab3c7c0102/gistfile1.txt

symphorien commented:
Can you try to remove the `stdenv.isLinux &&` guard on line 45 then ? This should solve ___cxa_end_catch at least and a bunch of others.

danieldk commented:
I have added

~~~
++ optional (stdenv.isDarwin && !withBundledLLVM) "-lc++"
~~~

Since the C++ standard library is libc++ on Darwin and the `--push-state` and `--as-needed` options are not known by LLVM ld (?). Now linking fails on missing jemalloc symbols, which is interesting, since configure reports that it finds jemalloc.

symphorien commented:
Yes `--push-state` is a GNU extension, I forgot.

danieldk commented:
Does the commented out comment about jemalloc prefixes still apply?

~~~
$ nm /nix/store/42lsjkifndxxzn4i8c3iyl0rijrsan0v-jemalloc-5.1.0/lib/libjemalloc.dylib  | grep calloc
00000000000045a0 T _calloc
00000000000023b0 T _je_bootstrap_calloc
000000000004a4b0 t _zone_calloc
~~~

symphorien commented:
Ah maybe, I removed it because it was not needed on linux.

danieldk commented:
That was indeed it, it finished the stage0 compiler, will keep you posted.

danieldk commented:
Here is the modified diff that compiles on Darwin:

https://gist.github.com/danieldk/127bf63c7ebcff0d7b0e9c21d482dd04

Changes:

* Add linker flag `-lc++` to `NIX_LDFLAGS` on Darwin.
* In postPatch, remove `je_` prefix in jemalloc symbol names.

Tested:

* nix-index derivation: succeeds
* fd derivation: fails with:
  ```
  error: process didn't exit successfully: `/private/tmp/nix-build-fd-7.2.0.drv-0/source/target/debug/deps/tests-090b3e427c9f00ad` (signal: 11, SIGSEGV: invalid memory reference)
  ```
* Some of my own crates (so far, so good).

This seems to solve the failing doctest problem.

danieldk commented:
The segmentation fault happens in `free()`:

```
* thread #2, name = 'test_absolute_path', stop reason = EXC_BAD_ACCESS (code=1, address=0x190178c63c)
  * frame #0: 0x00000001000c6b0b tests-090b3e427c9f00ad`free + 347
```

Full stack trace: https://gist.github.com/danieldk/5ee20615aa7ad61cba2a460ea5d08ac7

Mic92 commented:
So it crashes in jemalloc for some reason. What I am surprised about is that I do not see any jemalloc dependencies in the homebrew build: https://github.com/Homebrew/homebrew-core/blob/76dbba870f4e74d76392d992d96bd044daf435a9/Formula/rust.rb

@alyssais do you know if they install a jemalloc by default in homebrew?

symphorien commented:
There is a bundled jemalloc, which I remove to reduce compilation time. This was commented out before the PR. So maybe we should stick to the bundled jemalloc on darwin for the two or so upcoming rustc versions.
Afterwards, the build system will change wrt jemalloc because of https://github.com/rust-lang/rust/pull/55238 . At this point we may reevaluate unbundling jemalloc on darwin.

danieldk commented:
@Mic92 I think Rust imports jemalloc itself through jemalloc_sys.

I am now rebuilding jemalloc and rustc with the `je_` prefixes to see if that helps.

danieldk commented:
The `fd` unit tests pass without any segfaults when `jemalloc` is compiled with the `je_` prefix for jemalloc functions:

```
diff --git a/pkgs/development/libraries/jemalloc/common.nix b/pkgs/development/libraries/jemalloc/common.nix
index 593f4411f19..6cff824061e 100644
--- a/pkgs/development/libraries/jemalloc/common.nix
+++ b/pkgs/development/libraries/jemalloc/common.nix
@@ -12,7 +12,7 @@ stdenv.mkDerivation (rec {
   # By default, jemalloc puts a je_ prefix onto all its symbols on OSX, which
   # then stops downstream builds (mariadb in particular) from detecting it. This
   # option should remove the prefix and give us a working jemalloc.
-  configureFlags = stdenv.lib.optional stdenv.isDarwin "--with-jemalloc-prefix=";
+  #configureFlags = stdenv.lib.optional stdenv.isDarwin "--with-jemalloc-prefix=";
   doCheck = true;
 
   enableParallelBuilding = true;
```

From the jemalloc documentation:

> By default, the prefix is "", except on OS X, where it is "je_".  On OS X,
> jemalloc overlays the default malloc zone, but makes no attempt to actually
> replace the "malloc", "calloc", etc. symbols.

symphorien committed:
rustc: fix build with unbundled jemalloc and llvm on darwin

jemalloc with stripped prefix would cause segfaults in free:
https://github.com/NixOS/nixpkgs/pull/49557#issuecomment-436734677

Thanks to @danieldk for darwin testing/debugging.

symphorien commented:
Thanks @danieldk !
I rebased on a newer staging and added a commit taking into account what you found:
* add `-lc++` to `NIX_LDFLAGS`
* make jemalloc parametric to strip or not the `je_ `prefix and forcibly use the prefixed version on rustc

This new commit does not change hashes on linux.

danieldk commented:
Everything I tested before works well 👍 .

alyssais commented:
> @alyssais do you know if they install a jemalloc by default in homebrew?

Homebrew doesn't install anything "by default". The only things that'll get a Homebrew jemalloc are formulae that have `depends_on "jemalloc"` (`brew uses jemalloc`).

I don't [see](https://github.com/Homebrew/homebrew-core/search?q=jemalloc&unscoped_q=jemalloc) anything Rust or LLVM-related.

Mic92 commented:
aarch64 build is scheduled.

Mic92 commented:
aarch64 successfully builded `fd`

periklis commented:
@Mic92 can we enable tests for nix-index, alacritty and bat again?

periklis commented:
ic bat is enabled in #50236

alyssais commented:
> can we enable tests for nix-index, alacritty and bat again?

Only in staging, presumably.
